docs/content/guides/dagster/migrating-to-pythonic-resources-and-config.mdx

```diff
@@ -0,0 +1,295 @@
+from dagster import ConfigurableResource, asset
+
+class FancyDbResource(ConfigurableResource):
+    conn_string: str
+
+    def get_client(self) -> FancyDbClient:
+        return FancyDbClient(self.conn_string)
+
+@asset
+def new_asset(fancy_db: FancyDbResource) -> None:
+    client = fancy_db.get_client()
+    client.execute_query("SELECT * FROM foo")
+```
+
+`@resource` functions could also be context managers to handle initialization and cleanup tasks. This context manager was called managed by framework code rather than the user's code.
+
+```python file=/guides/dagster/migrating_to_python_resources_and_config/migrating_resources.py  startafter=begin_old_resource_code_contextmanager endbefore=end_old_resource_code_contextmanager dedent=4
+from dagster import InitResourceContext, asset, resource
+
+@resource(config_schema={"conn_string": str})
+def fancy_db_resource(context: InitResourceContext) -> Iterator[FancyDbClient]:
+    some_expensive_setup()
+    try:
+        # the client is yielded to the assets that require it
+        yield FancyDbClient(context.resource_config["conn_string"])
+    finally:
+        # this is only called once the asset has finished executing
+        some_expensive_teardown()
+
+@asset(required_resource_keys={"fancy_db"})
+def asset_one(context) -> None:
+    # some_expensive_setup() has been called, but some_expensive_teardown() has not
+    context.resources.fancy_db.execute_query("SELECT * FROM foo")
+```
+
+In the end, this caused confusion and difficult-to-understand stack traces. With new resources, the user manages this in the body of the asset or op directly.
+
+```python file=/guides/dagster/migrating_to_python_resources_and_config/migrating_resources.py  startafter=begin_new_resource_code_contextmanager endbefore=end_new_resource_code_contextmanager dedent=4
+from contextlib import contextmanager
+
+from dagster import ConfigurableResource, asset
+
+class FancyDbResource(ConfigurableResource):
+    conn_string: str
+
+    @contextmanager
+    def get_client(self) -> Iterator[FancyDbClient]:
+        try:
+            some_expensive_setup()
+            yield FancyDbClient(self.conn_string)
+        finally:
+            some_expensive_teardown()
+
+@asset
+def asset_one(fancy_db: FancyDbResource) -> None:
+    with fancy_db.get_client() as client:
+        client.execute_query("SELECT * FROM foo")
+```
+
+## Passing a resource directly while having the underlying client on the context
+
+This change to have the user manage accessing objects out of a resource rather than having the @resource factory function do it presents a problem with dealing with an at-scale migration: the old code will expect the client on the `context` object while the new code will except the enclosing resource object when it is accessed as a parameter.
```

Comment:
```suggestion
This change to have the user manage accessing objects out of a resource rather than having the `@resource` factory function do it presents a problem with dealing with an at-scale migration: the old code will expect the client on the `context` object while the new code will except the enclosing resource object when it is accessed as a parameter.
```